vault/seal/ocikms/ocikms.go

```diff
@@ -0,0 +1,414 @@
+	if authTypeAPIKeyStr != "" {
+		k.authTypeAPIKey, err = strconv.ParseBool(authTypeAPIKeyStr)
+		if err != nil {
+			metrics.IncrCounter(metricInitFailed, 1)
+			return nil, errwrap.Wrapf("failed parsing authTypeAPIKey parameter: {{err}}", err)
+		}
+	}
+	if k.authTypeAPIKey {
+		k.logger.Info("using OCI KMS with user principal")
+	} else {
+		k.logger.Info("using OCI KMS with instance principal")
+	}
+
+	// Check and set OCI KMS crypto client
+	if k.cryptoClient == nil {
+		kmsCryptoClient, err := k.getOCIKMSCryptoClient()
+		if err != nil {
+			metrics.IncrCounter(metricInitFailed, 1)
+			return nil, errwrap.Wrapf("error initializing OCI KMS client: {{err}}", err)
+		}
+
+		// KeyID validation by trying to generate a DEK
+		keyLength := 32
+		includePlaintextKey := true
+		generateKeyDetails := keymanagement.GenerateKeyDetails{
+			IncludePlaintextKey: &includePlaintextKey,
+			KeyId:               &k.keyID,
+			KeyShape: &keymanagement.KeyShape{
+				Algorithm: "AES",
+				Length:    &keyLength,
+			},
+		}
+		requestMetadata := k.getRequestMetadata()
+		dekInput := keymanagement.GenerateDataEncryptionKeyRequest{
+			GenerateKeyDetails: generateKeyDetails,
+			RequestMetadata:    requestMetadata,
+		}
+		generateDEKResponse, err := kmsCryptoClient.GenerateDataEncryptionKey(context.Background(), dekInput)
+		if err != nil || generateDEKResponse.Ciphertext == nil {
+			metrics.IncrCounter(metricInitFailed, 1)
+			return nil, errwrap.Wrapf("failed keyID validation: {{err}}", err)
+		}
+		k.logger.Info("successfully validated keyID")
+
+		// Store client
+		k.cryptoClient = kmsCryptoClient
+	}
+
+	// Check and set OCI KMS management client
+	if k.managementClient == nil {
+		kmsManagementClient, err := k.getOCIKMSManagementClient()
+		if err != nil {
+			metrics.IncrCounter(metricInitFailed, 1)
+			return nil, errwrap.Wrapf("error initializing OCI KMS client: {{err}}", err)
+		}
+
+		keyVersion, err := k.getCurrentKeyVersion(kmsManagementClient)
+		if err != nil {
+			metrics.IncrCounter(metricInitFailed, 1)
+			return nil, errwrap.Wrapf("error initializing OCI KMS client: {{err}}", err)
+		}
+		k.currentKeyID.Store(keyVersion)
+
+		k.managementClient = kmsManagementClient
+	}
+
+	// Map that holds non-sensitive configuration info
+	sealInfo := make(map[string]string)
+	sealInfo[KMSConfigKeyID] = k.keyID
+	sealInfo[KMSConfigCryptoEndpoint] = k.cryptoEndpoint
+	sealInfo[KMSConfigManagementEndpoint] = k.managementEndpoint
+
+	return sealInfo, nil
+}
+
+func (k *OCIKMSSeal) SealType() string {
+	return seal.OCIKMS
+}
+
+func (k *OCIKMSSeal) KeyID() string {
+	return k.currentKeyID.Load().(string)
+}
+
+func (k *OCIKMSSeal) Init(context.Context) error {
+	return nil
+}
+
+func (k *OCIKMSSeal) Finalize(context.Context) error {
+	return nil
+}
+
+func (k *OCIKMSSeal) Encrypt(ctx context.Context, plaintext []byte) (*physical.EncryptedBlobInfo, error) {
+	defer metrics.MeasureSince(metricEncrypt, time.Now())
+	if plaintext == nil || len(plaintext) == 0 {
```

Comment:
@chrishoffman I don't think the `len(plaintext)==0` is necessary, and was wondering if it can actually cause problems to error out here? Thinking if there is some seal-wrap edge case where 0 bytes are encrypted.